Cython/Utility/TypeConversion.c

```diff
@@ -584,15 +584,20 @@ static CYTHON_INLINE Py_UCS4 __Pyx_PyUnicode_AsPy_UCS4(PyObject*);
 /////////////// UnicodeAsUCS4 ///////////////
 
 static CYTHON_INLINE Py_UCS4 __Pyx_PyUnicode_AsPy_UCS4(PyObject* x) {
-   Py_ssize_t length;
-   length = PyUnicode_GET_LENGTH(x);
-   if (likely(length == 1)) {
-       return PyUnicode_READ_CHAR(x, 0);
-   }
-   PyErr_Format(PyExc_ValueError,
-                "only single character unicode strings can be converted to Py_UCS4, "
-                "got length %" CYTHON_FORMAT_SSIZE_T "d", length);
-   return (Py_UCS4)-1;
+    Py_ssize_t length;
+#if !CYTHON_COMPILING_IN_LIMITED_API
+    length = PyUnicode_GET_LENGTH(x);
+#else
+    length = PyUnicode_GetLength(x);
+    if (length < 0) return (Py_UCS4)-1;
+#endif
+    if (likely(length == 1)) {
+        return __Pyx_PyUnicode_READ_CHAR(x, 0);
+    }
+    PyErr_Format(PyExc_ValueError,
+                 "only single character unicode strings can be converted to Py_UCS4, "
+                 "got length %" CYTHON_FORMAT_SSIZE_T "d", length);
+    return (Py_UCS4)-1;
```

Comment:
This seems more of a case for `ASSUME_SAFE_SIZE`, but really, I'd just simplify it all together. The length checn is *very* unlikely to fail, and `!= 1` is an error either way.
```suggestion
    Py_ssize_t length = __Pyx_PyUnicode_GET_LENGTH(x);
    if (likely(length == 1)) {
        return __Pyx_PyUnicode_READ_CHAR(x, 0);
    } else if (likely(length >= 0)) {
        // "length == -1" indicates an error already.
        PyErr_Format(PyExc_ValueError,
                     "only single character unicode strings can be converted to Py_UCS4, "
                     "got length %" CYTHON_FORMAT_SSIZE_T "d", length);
    }
    return (Py_UCS4)-1;
```